build/codegen/codegen.sh

```diff
@@ -14,29 +14,61 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-GROUP_VERSIONS="ceph.rook.io:v1"
+set -o pipefail
+set -o nounset
+set -o errexit
+
+
+
+script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+base_dir="$(cd "${script_dir}/../.."  && pwd)"
+
+boilerplate="${base_dir}"/build/codegen/header.txt
+
+
+# echo "DEBUG: script_dir: ${script_dir}"
+# echo "DEBUG: CODE_GENERATOR: ${CODE_GENERATOR}"
+# echo "DEBUG: boilerplate ${boilerplate}"
+#  echo "DEBUG: done"
+# exit 0
+
+
+
+
+
+source "${CODE_GENERATOR}/kube_codegen.sh"
+
+
 
-scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 
 # CODE GENERATION
 # we run deepcopy and client,lister,informer generations separately so we can use the flag "--plural-exceptions"
 # which is only known by client,lister,informer binary and not the deepcopy binary
 
+
+
+(
+# because apis is a separate go module, code generation
+# must be run from within the corresponding directory.
+
+pushd ${base_dir}/pkg/apis
```

Comment:
```suggestion
(
  # because apis is a separate go module, code generation
  # must be run from within the corresponding directory.
  cd ${base_dir}/pkg/apis
```

Reply:
applied, conceptually.